CI: Link checker script

vipul-21 opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->
Enabling the link checker for the CI.
- There is know issue of warning: Circular inclusion in the include directive example: WARNING: circular inclusion in "include" directive. Not able to find the circular inclusion in the docs files. There is an open issue on the same : https://github.com/sphinx-doc/sphinx/issues/10951
- myst not able to parse the #anchor in the Documentation. example: 'myst' reference target not found. This has been disabled using the `ref.myst` in the suppress_warning for sphinx.
Note: This is dependent on merge of PR: https://github.com/cilium/cilium/issues/10601 to make sure CI passes


```release-note
ci: Enable link checker to ensure that all links in documentation are valid
```

Related: #10601

qmonnet commented on Documentation/check-build.sh:
```diff
@@ -153,13 +154,20 @@ else
+        echo "Link check failed!"
+        exit 1
+    fi
+fi
```
I think we should maybe just move this to a different script (say `Documentation/check-links.sh`), that we could call from `Documentation/Makefile`:

```makefile

linkcheck:
	@$(ECHO_CHECK) links
	$(QUIET)$(DOCKER_RUN) ./check-links.sh $(SPHINX_OPTS)
```

or directly from a new GitHub workflow, like we do for `check-build.sh` in `.github/workflows/documentation.yaml`.

vipul-21 replied:
I have made the script as a separate one.

vipul-21 replied:
I did try to run the github workflow but that error due permissions issue.
Do we want to keep that in the same PR ?

qmonnet replied:
Preferably in the same PR, so we can validate that the script works fine with the CI.

qmonnet reviewed (commented):
So we discussed about that at the community meeting, and the current consensus would be to do just like Tetragon does: a cron-task to check external links and to open GH issues when it finds any, but we would _not_ run the workflow on every PR.

So we need to have a way to run the link checker independently of the rest of the script. Let's maybe move it to a different script? (see below)

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,50 @@
+    if [[ -n "$MAKE_GIT_REPO_SAFE" ]]; then
+        git config --global --add safe.directory "${root_dir}"
+    fi
+    set -o nounset
```
The spell-checker needed this part because it runs some Git commands (to retrieve the list of authors and consider them as spelling exceptions), but I don't expect the linkchecker to do that, so we can probably remove this bit.

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,50 @@
+    rm -rf "${linksdir}"
+    
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
+     -q -E -w "${warnings}" 2>/dev/null
```
Nit: indent

```suggestion
    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
        -q -E -w "${warnings}" 2>/dev/null
```

qmonnet commented on Documentation/conf.py:
```diff
@@ -270,6 +270,31 @@
+linkcheck_anchors = False
+linkcheck_retries = 5
+
+tls_verify=False
```
```suggestion
tls_verify = False
```

Do you happen to know whether this option is used for anything else than the linkchecker? It's fine to disable for the linkchecker, but I wouldn't want to lower security if it were used for something else :thinking:

vipul-21 replied:
https://github.com/sphinx-doc/sphinx/pull/3156/files#diff-aac07aefec754086f5c344e2c233862f6a611c739b5647fd456c1b77d3577c79R10-R12, it is used for `intersphinx` as well

qmonnet replied:
They should really document that. Thanks a lot for looking it up! OK, intersphinx is not a concern and I don't see us using it anytime soon, so all good :+1:.

qmonnet reviewed (changes requested):
Hi, sorry for the delay (PTO). The separate script looks good, thanks! I just have some minor comments below. Ideally, could you squash all changes to `check-build.sh`/`check-links.sh` into a single commit (and keep the fixes for the broken links in a separate one)?

Ideally, yes I would prefer having the new script enable in the CI in this PR, so we can make sure the script works fine in CI.

vipul-21 committed:
fix: document links that were broken

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

vipul-21 committed:
ci: Adding the link checker as a script

Link checker is separated out from the check-build.sh into a separate
script. It creates a separate file with the broken links and filters
based on the known/expected errors. It is not added to the github
workflow as there is a known issue in the docutils pkg used.
(Issue: https://github.com/cilium/cilium/pull/27116#issuecomment-1737813302)

Co-authored-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: Vipul Singh <vipul21sept@gmail.com>

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,69 @@
+        run: |
+          set -ex
+          # chmod +x ./Documentation/build.sh
+          # ./Documentation/build.sh
```
We probably don't need to build the docs?

vipul-21 replied:
Hey sorry forgot to make it a draft. Was testing a bunch of things to make sure workflow works properly. Will ping you once I am done with the changes. 
Should be done by tomorrow

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,69 @@
+      - name: Check Documentation links
+        run: |
+          set -ex
+          cd Documentation && sudo make html && cd ..
```
Same here

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,69 @@
+          set -ex
+          cd Documentation && sudo make html && cd ..
+
+          chmod +x ./Documentation/check-links.sh
```
No, this should not be part of the workflow. You need to do it locally and then amend your commit to have the file with the updated permissions.

You can check locally the permissions for this file as known to Git:

```console
$ git ls-files -s Documentation/check-links.sh
100644 <SHA>
```

We want execution rights:

```console
$ chmod +x ./Documentation/check-links.sh
$ git add ./Documentation/check-links.sh
$ git commit --amend <REF>
```

Now we can check again then update the PR:

```console
$ git ls-files -s Documentation/check-links.sh
100755 <SHA>
```

vipul-21 replied:
Hey, thanks for this updated it.

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,69 @@
+          ./Documentation/check-links.sh
+          
+          ls ./Documentation/_build
+          sudo touch ./Documentation/_build/links_warnings.txt
```
This shouldn't be necessary? I take it you don't get any warnings in the file?

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,69 @@
+          
+          ls ./Documentation/_build
+          sudo touch ./Documentation/_build/links_warnings.txt
+          ./Documentation/_build/links_warnings.txt > $GITHUB_OUTPUT
```
```suggestion
          cat ./Documentation/_build/links_warnings.txt > $GITHUB_OUTPUT
```

qmonnet reviewed (commented):
Looking at the workflow code in passing, here are some observations in case it helps.

Anything in particular we can help with?

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,55 @@
+name: Periodically Check Documentation Links
```
```suggestion
name: Check Documentation Links
```

It's only periodical when we trigger it with the cron job; but the workflow itself is not necessarily periodical, especially if we can trigger it from PRs touching the workflow file. I'd remove this word from the title

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,55 @@
+          persist-credentials: false 
+      - name: Check Documentation links
+        run: |
+          ./Documentation/check-links.sh
```
I agree that the time elapsed for the test is not realistic. The script didn't run correctly, we could try a few things to check it:

- Run with `bash -x` to check where it exits
- Remove the `-q` option passed to Sphinx in `check-links.sh`. Actually, we should maybe do this even after we figured the issue. This would show that we effectively test the links, show how many rate-limit occur, or any warning that we skip, as well that all the redirects that happen.

My guess is simply that Sphinx (and the linkchecker) are not installed on the runner. We should run this command from the `docs-builder` container, which has all dependencies set up, just like we do for the documentation builds in CI (`documentation.yaml`):
```suggestion
      - name: Check Documentation links
        uses: docker://quay.io/cilium/docs-builder:4b194c0facc5d1d3b9c0dfef3529ee7c661a6b73@sha256:095cf9621cf46c6c878e25c3a98f02ef3d0aed1154936663e352fd3fe5e7613d
        with:
          entrypoint: ./Documentation/check-build.sh
```

vipul-21 replied:
I think this will require us to make changes to the script as well ?

qmonnet replied:
My bad, obviously I meant `check-links.sh` and not `check-build.sh`. I realised my mistake after having logged off yesterday but forgot to update in the evening :disappointed:.

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,43 @@
+    rm -rf "${linksdir}"
+    
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
+        -q -E -w "${warnings}" 2>/dev/null
```
```suggestion
        -E -w "${warnings}" 2>/dev/null
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,56 @@
+      - name: Check Documentation links
+        uses: docker://quay.io/cilium/docs-builder:4b194c0facc5d1d3b9c0dfef3529ee7c661a6b73@sha256:095cf9621cf46c6c878e25c3a98f02ef3d0aed1154936663e352fd3fe5e7613d
+        with:
+          entrypoint: ./Documentation/check-build.sh
```
```suggestion
          entrypoint: ./Documentation/check-links.sh
```

My bad, wrong suggestion.

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,43 @@
+    rm -rf "${linksdir}"
+    
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
+        -E -w "${warnings}" 2>/dev/null
```
```suggestion
        --color -E -w "${warnings}" 2>/dev/null
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
```
```suggestion
        uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+        with:
+          persist-credentials: false 
+      - name: Check Documentation links
+        uses: docker://quay.io/cilium/docs-builder:4b194c0facc5d1d3b9c0dfef3529ee7c661a6b73@sha256:095cf9621cf46c6c878e25c3a98f02ef3d0aed1154936663e352fd3fe5e7613d
```
```suggestion
        uses: docker://quay.io/cilium/docs-builder:9eb38bbec80e56a8a498b0cac42456e42cfe7f04@sha256:9936a2f648cb25cad6cf1fe5772ad1a5800d0033eb88b0dce77ad70a083661d9
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+          entrypoint: ./Documentation/check-links.sh
+      - name: Upload Documentation links report
+        if: ${{ failure() }}
+        uses: actions/upload-artifact@v3
```
```suggestion
        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+          issue_number=$(echo "$response" | jq -r '.items[0].number // empty')
+          echo "issue_number=$issue_number" >> $GITHUB_OUTPUT
+          echo $issue_number
+      - uses: actions/checkout@v3
```
```suggestion
      - uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+          echo $issue_number
+      - uses: actions/checkout@v3
+        name: Checkout
+      - uses: actions/download-artifact@v3
```
```suggestion
      - uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
```

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+          echo "issue_number=$issue_number" >> $GITHUB_OUTPUT
+          echo $issue_number
+      - uses: actions/checkout@v3
+        name: Checkout
```
Do we need to check out the code at this stage?

vipul-21 replied:
I guess not.

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,47 @@
+    rm -rf "${linksdir}"
+    
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
+        --color -E -q -w "${warnings}" -W --keep-going 2>/dev/null
```
```suggestion
        --color -E -w "${warnings}" -W --keep-going 2>/dev/null
```

qmonnet commented on Documentation/conf.py:
```diff
@@ -175,7 +175,7 @@
 
 # Ignore some warnings from MyST parser
-suppress_warnings = ['myst.header', 'myst.xref_missing']
+suppress_warnings = ["myst.header", "myst.xref_missing"]
```
This seems unrelated? Probably because this line was edited recently and your initial version would add the same warning exception? Let's keep the quotes, the file already uses a mix of both anyway.

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+        uses: peter-evans/close-issue@276d7966e389d888f011539a86c8920025ea0626 # v3.0.1
+        with:
+          issue-number: ${{ steps.search-issue.outputs.issue_number }}
+          comment: '[Periodic links check](https://github.com/cilium/cilium/actions/workflows/check-links-cron.yaml) no longer finds broken links, closing issue.'
```
Nit: There is no linebreak at the end of the file (this probably depends on your editor's configuration), can you please add one?

qmonnet reviewed (commented):
Looks mostly good, I've got some minor comments on Docker image pinning and a few other things. The link checker seems to run well, now we need to validate the issue creation on broken link detection.

> The checks for creating/updating the issue needs the write permission. I don't have the permission to assign that to the github workflow stage.

I think I should be able to check if I open another PR from the cilium repository? I'll give it a try.

qmonnet commented on .github/workflows/check-links-cron.yaml:
```diff
@@ -0,0 +1,71 @@
+        name: Checkout
+      - uses: actions/download-artifact@v3
+        name: Download Documentation links report
+        if: ${{ failure() && needs.links-checker.outputs.status == 'failure' }}
```
While the `failure()` here? Doesn't it mean that we run the step only if the previous step (checking out the code) fails? It seems to have prevented the download from happening on this test: https://github.com/cilium/cilium/actions/runs/6314074348/job/17143831502?pr=28284

vipul-21 replied:
Isn't the `failure` suppose to check if any of the steps fails of job ? An in case of chain, the dependent job's steps are also considered.

qmonnet replied:
> Isn't the `failure` suppose to check if any of the steps fails of job ?

Honestly I'm not familiar enough to tell. All I know is something prevented the artifact download to run on my test. Let me try another time with your updated version

qmonnet commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,47 @@
+build_with_linkchecker() {
+    rm -rf "${linksdir}"
+    
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
```
Nit: trailing spaces

```suggestion
    rm -rf "${linksdir}"

    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
```

qmonnet reviewed (changes requested):
Looking at [my latest run](https://github.com/cilium/cilium/actions/runs/6327248302/job/17183023483), it was maybe not the `failure()` that blocked the artifact download, but possibly the `needs.links-checker.outputs.status == 'failure'` instead: I don't think we're ever setting this output for `links-checker`? Looking at a similar usage of `outputs.status` in file `.github/workflows/conformance-ginkgo.yaml`, there is `echo "status=failure" >> $GITHUB_OUTPUT` to set the status, it's not automatic. So we probably need something similar.

Annoyingly, the link checker step also failed (consistently, it seems) on this new run, and I've no idea what's causing it:

```
[...]
reading sources... [ 58%] gettingstarted/hubble_setup
reading sources... [ 58%] gettingstarted/k8s-install-default
Please fix the following documentation warnings:

Exception occurred:
  File "/usr/local/lib/python3.11/site-packages/docutils/parsers/rst/states.py", line 2297, in comment
    self.document.include_log.pop()
IndexError: pop from empty list
The full traceback has been saved in /tmp/sphinx-err-1xnht2gf.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!

Adding the warnings to the file.
```

Did you observe this before, by any chance?

qmonnet reviewed (changes requested):
The changes look all good, thanks a lot.

On the second commit, can you please:

- Update the commit title, given that we no longer enable the link checker in the CI
- Add a commit description that summarises what we're doing here (and why we don't enable it in CI)
- Drop the `Co-authored-by:` tag, or updated it with my work email address (`<quentin@isovalent.com>`)

We should be all good after that

qmonnet reviewed (approved):
Thanks a lot for this work!

joestringer commented on Documentation/check-links.sh:
```diff
@@ -0,0 +1,48 @@
+    rm -rf "${linksdir}"
+
+    sphinx-build -b linkcheck -d "${build_dir}/doctrees" . "${linksdir}" \
+        --color -E -w "${warnings}" -W --keep-going 2>/dev/null
```
Why is stderr being redirected here? Could this hide real errors?

qmonnet replied:
```console
$ sphinx-build --help
[...]
  -w FILE           write warnings (and errors) to given file
```

We keep (and print, when relevant) both warnings and errors in the provided file, so I think it's safe to ignore `stderr`. If we keep it, it makes duplicate entries with the file that we dump at the end. One motivation for dumping the file instead of simply using `stderr` is that the errors and warnings can easily get lost in the middle of the info-level output.
